spec/justified.tla

```diff
@@ -0,0 +1,209 @@
+\* VotesRelevantForOneStepIteration(C_G, ...) = {}, eventually S_m will be empty.
+
+\* This satisfies our termination requirement from the recursion rule.
+
+\* ============IMPORTANT============
+\* TODO: Instead of computing is_justified_checkpoint for a single checkpoint
+\* we should directly define the set of _ALL_ justified checkpoints for a given state
+\* (which can be justified by all the votes in that state)
+\* =================================
+
+\* Using the above knowledge, we can construct a Chain operator, to compute the sequence <<M_n, ..., M_1>>,
+\* as required by the recursion rule.
+
+\* @typeAlias: targetMap = $checkpoint -> Set($signedVoteMessage);
+\* @type: ($targetMap, Int, $commonNodeState) => Seq($targetMap);
+Chain(x, N, node_state) ==
+    LET 
+        \* @type: ($targetMap) => Bool;
+        P(map) == DOMAIN map = {}
+        \* if x = M_i, then b(x) defines M_{i+1}
+        \* @type: ($targetMap) => $targetMap;
+        b(map) ==
+            LET newTargets == UNION { Sources(x[target]): target \in DOMAIN map } 
+            IN [ target \in newTargets |-> VotesRelevantForOneStepIteration(target, node_state) ]
+    IN  LET 
+            \* @type: (Seq($targetMap), Int) => Seq($targetMap);
+            step(seq, i) == 
+                IF P(seq[1])
+                THEN seq
+                ELSE <<b(seq[1])>> \o seq \* Alternatively, we can append here and reverse the list at the end
+        IN ApaFoldSeqLeft( step, <<x>>, MkSeq(N, LAMBDA i: i) )
+        
+
+\* @type: ($targetMap, Int, $commonNodeState) => Set($checkpoint);
+AllJustifiedCheckpoints(initialTargetMap, N, node_state) ==
+    LET chain == Chain(initialTargetMap, N, node_state) IN
+    LET 
+        \* @type: ($targetMap, Set($checkpoint)) => Set($checkpoint);
+        G(currentTargetMap, justifiedCheckpoints) ==
+            LET targets == DOMAIN currentTargetMap IN
+            \* We define the justification filter for one checkpoint, which
+            \* looks at the currently computed set of justifiedCheckpoints
+            \* instead of recursing
+            LET 
+                \* @type: ($checkpoint) => Bool;
+                isOneCheckpointJustified(checkpoint) ==
+                    LET
+                        \* TODO: If we end up always doing 1-state model checking, these two are
+                        \* implied by the sate validity predicate and can be omitted
+                        hasBlockHash == has_block_hash(checkpoint.block_hash, node_state)
+                        isCompleteChain == 
+                            is_complete_chain(
+                                get_block_from_hash(checkpoint.block_hash, node_state), 
+                                node_state
+                            )
+                    IN  LET 
+                        \* TODO: Since GET_VALIDATOR_SET_FOR_SLOT is a constant operator, we could
+                        \* directly substitute it here
```

Comment:
I think we'll want to play with parameters for `GET_VALIDATOR_SET_FOR_SLOT` later

Reply:
I'll open an issue, since this part requires some back-and-forth with our collaborators.